# codon-bias

This package provides codon usage bias (CUB) analysis tools for genomic sequences, focusing on protein coding regions, translation efficiency and synonymous mutations. These include implementations of popular models from the past four decades of codon usage study, such as:

- Nucleotide and codon k-mer statistics (GC, GC3, CpG, etc.)
- Frequency of Optimal Codons (FOP)
- Relative Synonymous Codon Usage (RSCU)
- Codon Adaptation Index (CAI), including extensions:
    - Codon pair (and k-mers) adaptation
- Effective Number of Codons (ENC), including extensions:
    - Background correction
    - Improved estimation
    - Effective number of codon pairs (and k-mers) (ENcp)
- tRNA Adaptation Index (tAI)
- Codon Pair Bias (CPB/CPS)
- Relative Codon Bias Score (RCBS)
- Normalized Translational Efficiency (nTE)
- Directional Codon Bias Score (DCBS)
- Codon Usage Frequency Similarity (CUFS)

This package also includes tools for sequence optimization based on these codon usage models, and generators of random sequence permutations that can be used to compute empirical p-values and z-scores.

## installation

```
pip install codon-bias
```

## documentation

Read on [Read the Docs](https://codon-bias.readthedocs.org).

## contributing

Contributions of additional models to the package are welcome! Please familiarize yourself with the existing classes, and try to conform to their style.
